### PR TITLE
Adding TRT-LLM template

### PR DESCRIPTION
**Pull request by aspctu**
## :rocket: What

This PR adds support for a generic TRT-LLM + Triton template. Most of our TRT-LLM trusses in [truss examples](https://github.com/basetenlabs/truss-examples/) have the same code hence this PR. 

This PR also adds support for config-driven engine builds at runtime. See below for an example `config.yaml` that builds and serves an engine at run-time. 

## :computer: How

We create a seperate template for the TRT-LLM truss. We create a traditional truss context and then overwrite the `model` and `packages` directory with the TRT-LLM truss. The TRT-LLM truss contains all the logic for building and serving engines with TensorRT-LLM and TritonServer. 

## :microscope: Testing

Tested locally with a couple sample configurations.

### Appendix
    Example (for building an engine)
    ```
    build:
        model_server: TRT_LLM
        arguments:
            tokenizer_repository: "mistralai/mistral-v2-instruct"
            arguments:
                max_input_len: 1024
                max_output_len: 1024
                max_batch_size: 64
            quant: "weights_kv_int8"
            tensor_parallel_count: 2
            pipeline_parallel_count: 1
    ```

    Example (for using an existing engine)
    ```
    build:
        model_server: TRT_LLM
        arguments:
            engine_repository: "baseten/mistral-v2-32k"
            tensor_parallel_count: 2
            pipeline_parallel_count: 1
    ```

### CODE REVIEW

**Review by joostinyi — APPROVED** [review]
LGTM. I don't see `build.model_server` in the example config that you attached so I would just make sure that works as intended. Also +1 for consolidating the appropriate build + serving metadata, I feel like `metadata.engine`, `metadata.tensor_parallelism` (this one can just be removed in favor of `tp_size`), and `metadata.tokenizer_repository` can all be moved under the `build` kwarg and be read from there as a single source of truth

File: truss/templates/trtllm/packages/build_engine_utils.py
Comment (marius-baseten) [inline truss/templates/trtllm/packages/build_engine_utils.py:57]: Nit: consider `=`-style flags. I personally find it easier to visually parse into arg_name arg_value pairs than spaces...

File: truss/templates/trtllm/packages/build_engine_utils.py
Comment (marius-baseten) [inline truss/templates/trtllm/packages/build_engine_utils.py:53]: Nit: add return type.

File: truss/templates/trtllm/packages/build_engine_utils.py
Comment (marius-baseten) [inline truss/templates/trtllm/packages/build_engine_utils.py:88]: Is this needed? Can't pydantic parse such a dict directly?

File: truss/templates/trtllm/packages/build_engine_utils.py
Comment (marius-baseten) [inline truss/templates/trtllm/packages/build_engine_utils.py:115]: Is `Engine` an actual engine or just config - does the instantiation of this class already execute the building process?

**Review by marius-baseten — APPROVED** [review]

File: truss/templates/trtllm/packages/utils.py
Comment (aspctu) [reply truss/templates/trtllm/packages/utils.py:6]: In this example, you propose we instead import 
```
import tritonclient.grpc as grpcclient
import huggingface_hub
import tritonclient
```

and use the symbols as:
```
huggingface_hub.snapshot_download(...)
tritonclient.utils.np_to_triton_dtype(...)
```
?

File: truss/contexts/image_builder/serving_image_builder.py
Comment (marius-baseten) [reply truss/contexts/image_builder/serving_image_builder.py]: FYI: truss/errors.py

File: truss/templates/trtllm/packages/utils.py
Comment (aspctu) [reply truss/templates/trtllm/packages/utils.py:14]: this is a good point, i think we can get away with just using `shutil.copy_tree` here

File: truss/templates/trtllm/packages/client.py
Comment (marius-baseten) [inline truss/templates/trtllm/packages/client.py:55]: I just happened to got hung up with this while playing with spec dec which does not have an ensemble. Can me find a condition that is not depending on a specific model name? Maybe `is_server_ready()`?

File: truss/contexts/image_builder/serving_image_builder.py
Comment (aspctu) [reply truss/contexts/image_builder/serving_image_builder.py:528]: noted, can add later when we have more errors to define.

File: truss/templates/trtllm/packages/build_engine_utils.py
Comment (aspctu) [reply truss/templates/trtllm/packages/build_engine_utils.py:21]: `Engine` is a config, we use that to then kick off the building of an actual Engine.

File: truss/templates/trtllm/packages/build_engine_utils.py
Comment (joostinyi) [inline truss/templates/trtllm/packages/build_engine_utils.py:24]: Let's make a TODO to change `build_engine` to be able to build the engine directly to the destination path to avoid copying over these files

File: truss/templates/trtllm/packages/utils.py
Comment (joostinyi) [reply truss/templates/trtllm/packages/utils.py:16]: Is there additional context here for why we are choosing not to use `shutil.move()`? Are we going to address this in a follow up?

File: truss/templates/trtllm/packages/utils.py
Comment (joostinyi) [reply truss/templates/trtllm/packages/utils.py:49]: https://github.com/huggingface/huggingface_hub/blob/8edc3f47900e5a980865b739681143086211a630/src/huggingface_hub/_snapshot_download.py#L65 has good context for the cases here

File: truss/templates/trtllm/packages/constants.py
Comment (joostinyi) [inline truss/templates/trtllm/packages/constants.py:3]: Can we make a note to parametrize and render these pbtxt at truss load time as with a pattern like in the TRT-LLM backend [tooling](https://github.com/triton-inference-server/tensorrtllm_backend/blob/7982723f13ddad4a478093fea02f55441b3608b0/tools/fill_template.py)?

File: truss/templates/trtllm/packages/build_engine_utils.py
Comment (joostinyi) [reply truss/templates/trtllm/packages/build_engine_utils.py:21]: Let's make a follow up ticket to clean up this class name to something less ambiguous so we know this is the configuration used to build an engine from

**Review by joostinyi — APPROVED** [review]

File: truss/templates/trtllm/packages/build_engine_utils.py
Comment (marius-baseten) [inline truss/templates/trtllm/packages/build_engine_utils.py:21]: If engine is a pydantic model, it's more idiomatic to use the parse functions (V1 `parse_obj` V2 `model_validate`).

But I'm confused, truss repo uses `pydantic = "^1.10.9"` and `model_dump` is from the V2 API.

https://github.com/basetenlabs/truss/blob/main/pyproject.toml#L35C1-L35C9

File: truss/templates/trtllm/packages/build_engine_utils.py
Comment (marius-baseten) [inline truss/templates/trtllm/packages/build_engine_utils.py:17]: Can you pull `sys.path.append("/app/baseten")` down here to make it also explained by this note (espcially since the hardcoded path literal is assuming a very particular execution context).
And maybe make it actually a TODO, since I think we should long term find a more principled solution for this.

File: truss/templates/trtllm/model/model.py
Comment (aspctu) [reply truss/templates/trtllm/model/model.py:20]: Truss always passes these variables, so we don't have to worry about managing their existence.

Comment (marius-baseten) [reply truss/templates/trtllm/model/model.py:20]: Right, then why not make an explicit function signature instead of `**kwargs`?

File: truss/templates/trtllm/packages/schema.py
Comment (aspctu) [reply truss/templates/trtllm/packages/schema.py:78]: I agree, this change will need to be implemented first on the build engine logic. I'll hold off on this and make tickets to update that codebase (a couple issues here require changes there before implementing the interface here).

File: truss/templates/trtllm/packages/schema.py
Comment (aspctu) [reply truss/templates/trtllm/packages/schema.py:90]: I agree, this change will need to be implemented first on the build engine logic. I'll hold off on this and make tickets to update that codebase (a couple issues here require changes there before implementing the interface here).

File: truss/templates/trtllm/packages/schema.py
Comment (aspctu) [reply truss/templates/trtllm/packages/schema.py:149]: They're split out as part of the build logic. We can update the build logic, and then update the interface here.

File: truss/templates/trtllm/model/model.py
Comment (aspctu) [reply truss/templates/trtllm/model/model.py:59]: The line before this, where we call the Triton Server to `start` blocks until the server is live and ready so a) isn't an issue as well.

File: truss/templates/trtllm/model/model.py
Comment (aspctu) [reply truss/templates/trtllm/model/model.py:13]: Spoke offline about this, broader discussion needed about enforcing code styles. Will defer to follow-up for now.

File: truss/templates/trtllm/packages/build_engine_utils.py
Comment (aspctu) [reply truss/templates/trtllm/packages/build_engine_utils.py:20]: I think this is another artifact of the base image used. The base image has Pydantic installed. Also, that pyproject requirements are not used at run-time. What's used at run-time is this file:
https://github.com/basetenlabs/truss/blob/main/truss/templates/server/requirements.txt

**Pull request merged**
